src/journal/compress.c

```diff
@@ -667,13 +677,229 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
         return -EPROTONOSUPPORT;
 #endif
 }
+#if HAVE_ZSTD
+static int zstd_ret_to_errno(size_t ret) {
+        static_assert(ZSTD_VERSION_NUMBER >= 10301, "<zstd_errors.h> is stable since ZSTD 1.3.1");
+
+        ZSTD_ErrorCode error = ZSTD_getErrorCode(ret);
+
+        switch (error) {
+        case ZSTD_error_dstSize_tooSmall:
+                return -ENOBUFS;
+        case ZSTD_error_memory_allocation:
+                return -ENOMEM;
+        default:
+                return -EBADMSG;
+        }
+}
+#endif
+
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *cctx = NULL;
+        _cleanup_free_ void *in_buff = NULL, *out_buff = NULL;
+        size_t in_allocsize, out_allocsize;
+        size_t z;
+        uint64_t left = max_bytes, in_bytes = 0;
+        /* This can be used in the future to add uncompressed size to the header */
+        uint64_t in_totalsize = 0;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        /* Create the context and buffers */
+        in_allocsize = ZSTD_CStreamInSize();
+        out_allocsize = ZSTD_CStreamOutSize();
+        in_buff = malloc(in_allocsize);
+        out_buff = malloc(out_allocsize);
+        cctx = ZSTD_createCCtx();
+        if (!cctx || !out_buff || !in_buff)
+                return -ENOMEM;
+
+        if (in_totalsize) {
+                z = ZSTD_CCtx_setPledgedSrcSize(cctx, in_totalsize);
+                if (z)
+                        log_debug("Failed to enable ZSTD input size, ignoring: %s", ZSTD_getErrorName(z));
+        }
+
+        z = ZSTD_CCtx_setParameter(cctx, ZSTD_c_checksumFlag, 1);
+        if (z)
+                log_debug("Failed to enable ZSTD checksum, ignoring: %s", ZSTD_getErrorName(z));
+
+        /* This loop read from the input file, compresses that entire chunk,
+         * and writes all output produced to the output file.
+         */
+        for (bool is_last_chunk = false; !is_last_chunk;) {
```

Comment:
i must say, this creative use of for() loops doesn't really help readability... we write code that is suppsoed to be easy to read, and not ooze out creative uses of C wherever possible... ;-)

hence, i think this is more readable:

```
bool is_last_chunk = false;
do {
…
while (!is_last_chunk);
```

I am pretty sure most people would find this a lot close to what they would expect for a loop until some condition is true. Let's leave for() as something we use for iterating...

Reply:
except the main loop condition is `while red > 0`, changed `is_last_chunk` to break out of an infinite loop.